### PR TITLE
MicroFocus Contributed - Backward Compatibility with AccuRev Versions for improved efficiency

### PR DESCRIPTION
**Pull request by PoojaVishal**
Hist command with combined transaction types where not supported in lower version than 7.X.
Now efficiency improved code only if AccuRev is 7 or higher and same old performance code in lower versions.
We hit accurev executable command to get version and parse it with new parser ParseAccrevVersion.

https://github.com/jenkinsci/accurev-plugin/pull/75

Closed 75 PR because i created it on master.

### CODE REVIEW

File: src/main/java/hudson/plugins/accurev/delegates/StreamDelegate.java [hunk 1]
```diff
@@ -88,6 +91,31 @@ protected PollingResult checkForChanges(Job<?, ?> project)
           .println("Tried to find '" + localStream + "' Stream, could not found it.");
       return PollingResult.NO_CHANGES;
     }
+    // command to get version of accurev
+    final ArgumentListBuilder cmd = new ArgumentListBuilder();
+    cmd.add("accurev");
+    String ACCUREV_VERSION =
+        AccurevLauncher.runJustAccurev(
```
Comment (casz) [inline src/main/java/hudson/plugins/accurev/delegates/StreamDelegate.java]: I doubt you need to create a separate method for the Launcher.

You should be able to call `AccurevLauncher.runCommand()` with an empty argumentList `final ArgumentListBuilder cmd = new ArgumentListBuilder()`

File: src/main/java/hudson/plugins/accurev/delegates/StreamDelegate.java [hunk 2]
```diff
@@ -88,6 +91,31 @@ protected PollingResult checkForChanges(Job<?, ?> project)
           .println("Tried to find '" + localStream + "' Stream, could not found it.");
       return PollingResult.NO_CHANGES;
     }
+    // command to get version of accurev
+    final ArgumentListBuilder cmd = new ArgumentListBuilder();
+    cmd.add("accurev");
```
Comment (casz) [inline src/main/java/hudson/plugins/accurev/delegates/StreamDelegate.java]:
```suggestion
```

File: src/main/java/hudson/plugins/accurev/parsers/output/ParseAccuRevVersion.java [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+import hudson.plugins.accurev.AccurevLauncher.ICmdOutputParser;
+import hudson.plugins.accurev.AccurevLauncher.UnhandledAccurevCommandOutput;
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.io.Reader;
+import java.nio.charset.Charset;
+
+public final class ParseAccuRevVersion implements ICmdOutputParser<String, Void> {
```
Comment (casz) [inline src/main/java/hudson/plugins/accurev/parsers/output/ParseAccuRevVersion.java]: Hmm you already have command called `getAccuRevVersion` however that does not support getting the path to AccuRev

https://github.com/jenkinsci/accurev-plugin/blob/e01b1affe0b3fb648a53b4928907e366306cd025/src/main/java/hudson/plugins/accurev/cmd/GetAccuRevVersion.java

File: src/main/java/hudson/plugins/accurev/AccurevLauncher.java [hunk 4]
```diff
@@ -383,6 +383,60 @@ public static boolean runCommand( //
     }
   }
 
+  public static <TResult, TContext> TResult runJustAccurev(
```
Comment (casz) [inline src/main/java/hudson/plugins/accurev/AccurevLauncher.java]: Pretty sure this method is not necessary 👍

File: src/main/java/hudson/plugins/accurev/parsers/output/ParseAccuRevVersion.java (hunk 3 above)
Comment (PoojaVishal) [reply src/main/java/hudson/plugins/accurev/parsers/output/ParseAccuRevVersion.java]: Yea.. and in linux it gives issues.. i tried that already.

File: src/main/java/hudson/plugins/accurev/delegates/StreamDelegate.java (hunk 1 above)
Comment (PoojaVishal) [reply src/main/java/hudson/plugins/accurev/delegates/StreamDelegate.java]: Ok i will try that i push changes asap. Got to get merged today

File: src/main/java/hudson/plugins/accurev/delegates/StreamDelegate.java (hunk 2 above)
Comment (PoojaVishal) [reply src/main/java/hudson/plugins/accurev/delegates/StreamDelegate.java]: if (!accurevPath.equals(machineReadableCommand.toCommandArray()[0])) {
      machineReadableCommand.prepend(accurevPath);
    }
this line launcher run command may give nullpointer na?

File: src/main/java/hudson/plugins/accurev/parsers/output/ParseAccuRevVersion.java (hunk 3 above)
Comment (casz) [reply src/main/java/hudson/plugins/accurev/parsers/output/ParseAccuRevVersion.java]: Please replace `GetAccuRevVersion` with the new version :)

File: src/main/java/hudson/plugins/accurev/delegates/AbstractModeDelegate.java [hunk 5]
```diff
@@ -546,20 +562,18 @@ protected void buildEnvVarsCustom(Run<?, ?> build, Map<String, String> env) {
    * color parameter will be style if version greater than 6 like 6.1.x or 7 then color parameter
    * will streamStyle
    *
+   * @param version
    * @return String
    */
-  private String getStreamTypeParameter() {
-    String fullVersion = GetAccuRevVersion.getAccuRevVersion().trim();
```
[the remark this answers is not in the document]
Comment (PoojaVishal) [reply src/main/java/hudson/plugins/accurev/delegates/AbstractModeDelegate.java]: Done. But still its not reflecting here.
image

**Comment by PoojaVishal** [comment]
https://github.com/jenkinsci/accurev-plugin/pull/74
will this pull request impact my changes in anyway?

**Comment by PoojaVishal** [comment]
@casz  please merge my changes.

**Comment by casz** [comment]
I think you forgot to push the remove unused class :)

https://github.com/jenkinsci/accurev-plugin/pull/76/commits

**Comment by PoojaVishal** [comment]
@casz see now

**Comment by PoojaVishal** [comment]
@can you please merge my change now?

**Comment by PoojaVishal** [comment]
@casz  will that 74 PR effect in anyway?

**Comment by casz** [comment]
Not sure what you mean :)

**Comment by PoojaVishal** [comment]
https://github.com/jenkinsci/accurev-plugin/pull/74
this will affect my changes?

**Comment by casz** [comment]
Still not sure what you mean 😕 

They are both merged and they both build successfully. So it other words no?

**Comment by PoojaVishal** [comment]
@casz build failed

**Comment by PoojaVishal** [comment]
@casz  can you please see whats wrong..
need the final build.

**Comment by PoojaVishal** [comment]
image

It failed after build success.

**Comment by casz** [comment]
Ya, not too worried ci.jenkins.io trobuled at times.

**Comment by PoojaVishal** [comment]
thank you so much @casz 
Hope build is success now.

**Comment by PoojaVishal** [comment]
@casz, all builds are stopped at same point.. finding maven. Build Stil running..

**Comment by casz** [comment]
Added GitHub Action CI which built successfully 
https://github.com/jenkinsci/accurev-plugin/runs/265905924

I wouldn't worry too much about ci.jenkins.io it's going through some teething issues.

**Comment by casz** [comment]
FYI I already released a new version: https://github.com/jenkinsci/accurev-plugin/releases/tag/accurev-0.7.21

**Pull request merged**